engine/kafka/src/main/scala/pl/touk/nussknacker/engine/kafka/validator/TopicsExistenceValidator.scala

```diff
@@ -0,0 +1,107 @@
+
+import scala.collection.JavaConverters._
+
+trait TopicsExistenceValidator {
+
+  final def validateTopic(topic: String): Validated[TopicExistenceValidationException, String] = validateTopics(List(topic)).map(_.head)
+
+  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]]
+}
+
+trait WithCachedTopicsExistenceValidator extends TopicsExistenceValidator {
+  protected val kafkaConfig: KafkaConfig
+  protected val validatorConfig: CachedTopicsExistenceValidatorConfig = DefaultConfig
+  private lazy val validator = new CachedTopicsExistenceValidator(kafkaConfig, validatorConfig)
+
+  final override def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] = validator.validateTopics(topics)
+}
+
+case class CachedTopicsExistenceValidatorConfig(autoCreateFlagFetchCacheTtl: FiniteDuration, topicsFetchCacheTtl: FiniteDuration, adminClientTimeout: FiniteDuration)
+
+object CachedTopicsExistenceValidatorConfig {
+  val AutoCreateTopicPropertyName = "auto.create.topics.enable"
+  val TopicExistenceValidationEnabledPropertyName = "topic.existence.validation.enabled"
+  val ValidationEnabledDefault = false
+  val DefaultConfig = CachedTopicsExistenceValidatorConfig(
+    autoCreateFlagFetchCacheTtl = 5 minutes, topicsFetchCacheTtl = 30 seconds, adminClientTimeout = 500 millis
+  )
+}
+
+class CachedTopicsExistenceValidator(kafkaConfig: KafkaConfig, config: CachedTopicsExistenceValidatorConfig)
+  extends TopicsExistenceValidator with LazyLogging {
+
+  private lazy val autoCreateSettingCache = new SingleValueCache[Boolean](expireAfterAccess = None, expireAfterWrite = Some(config.autoCreateFlagFetchCacheTtl))
+  private lazy val topicListCache = new SingleValueCache[Set[String]](expireAfterAccess = None, expireAfterWrite = Some(config.topicsFetchCacheTtl))
+
+  private lazy val adminClientProperties = {
+    val properties = new Properties()
+    properties.setProperty("bootstrap.servers", kafkaConfig.kafkaAddress)
+    properties
+  }
+
+  def validateTopics(topics: List[String]): Validated[TopicExistenceValidationException, List[String]] = {
+    if (validationDisabled || isAutoCreateEnabled) {
+      Valid(topics)
+    } else {
+      val existingTopics = topicListCache.getOrCreate {
+        usingAdminClient {
+          _.listTopics(new ListTopicsOptions().timeoutMs(config.adminClientTimeout.toMillis.toInt))
+            .names().get().asScala.toSet
+        }
+      }
+      val notExistingTopics = topics.diff(existingTopics.toList)
+
+      if (notExistingTopics.isEmpty)
+        Valid(topics)
+      else
+        Invalid(new TopicExistenceValidationException(notExistingTopics))
+    }
+  }
+
+  def validationDisabled: Boolean = !kafkaConfig.kafkaEspProperties.flatMap(_.get(TopicExistenceValidationEnabledPropertyName))
+    .map(_.toBoolean).getOrElse(ValidationEnabledDefault)
+
+  private def isAutoCreateEnabled: Boolean = autoCreateSettingCache.getOrCreate {
```

Comment:
We'd rather have a test for that? this code looks a bit quirky - and I'd suspect it may be a bit more kafka version dependent than most of other codes

Reply:
I've added integration test for this validator, will it be enough?

Reply:
:+1: